### PR TITLE
allocation hotspot in small Map iterators

### PR DESCRIPTION
**Pull request by mkeskells**
avoid vararg call to generic Iterator apply

in a similar way to the way Set1-4 work

### CODE REVIEW

**Comment by retronym** [comment]
Needs a `ProblemFilters.exclude[MissingClassProblem]("scala.collection.immutable.Map$MapNIterator")` (test with `sbt library/mimaReportBinaryIssues`)

**Review by diesalbla — APPROVED** [review]

File: src/library/scala/collection/immutable/Map.scala [hunk 1]
```diff
@@ -136,6 +136,28 @@ object Map extends ImmutableMapFactory[Map] {
     override def hashCode: Int = MurmurHash3.emptyMapHash
     override private[immutable] def foreachEntry[U](f: (Any, Nothing) => U): Unit = ()
   }
+  @SerialVersionUID(3L)
+  private abstract class MapNIterator[K,V](n: Int) extends AbstractIterator[(K,V)] with Serializable {
+    private[this] var current = 0
```
Comment (javax-swing) [inline src/library/scala/collection/immutable/Map.scala:143]: seems a bit unnecessary to keep track of current and remainder

**Review by javax-swing — COMMENTED** [review]
are there any benchmarks that show that this fixes the allocation problem in hotspot?

File: src/library/scala/collection/immutable/Map.scala (hunk 1 above)
Comment (mkeskells) [reply src/library/scala/collection/immutable/Map.scala:143]: agreed

**Comment by mkeskells** [comment]
reworked to include allocation tests and cover `keysIterator` and `valuesIterator`

**Review by retronym — APPROVED** [review]

**Pull request merged**